### ISSUE TITLE
Make the `Frame` struct `Pixel`-agnostic?

### ISSUE THREAD

**Issue by moulins** [image-rs/image#2398]
The `Frame`s returned by `AnimationDecoder` are hardcoded to contain `Rgba8Image`s, which is generally suboptimal when working with animated PNG or WEBP images:
- if the source image has no alpha channel, this imposes an extra conversion;
- 16-bit APNGs are unsupported because of this.

Would it make sense to have `Frame` contain a `DynamicImage` instead of a `Rgba8Image`, despite the fact that this is a breaking change? Existing users of the API can easily upgrade by doing the conversion to Rgba8 manually with `.into_rgba8()` or similar.

**Comment by kornelski** [comment]
It would be nice to have. 

It might be doable in a backwards-compatible way by adding a new method next to `into_frames`, like `into_frames_dynamic()`.

However, I don't think it has practical utility besides optimization for opaque RGB8. AFAIK WebP doesn't have 16-bit coding — lossless is 8-bit-only, and VP8 implementations don't even use full range in 8 bits. PNG compression is pretty bad in 16 bits.

**Comment by moulins** [comment]
You're right that this would be mostly for optimization purposes; another theoretical use-case would be grayscale APNGs, but I'd be surprised if they actually exist in the wild ^^

**Comment by 197g** [comment]
For unsupported color layouts, #2522 should be the general direction. It's probably _fine_ to convert the color while encoding. At least truncating the depth was a sought after compatibility adjustment to be made. Color space treatment such as rgba->luma might be a different beast but it's not a likely problem for most common formats. Event though some formats can not encode a CICP indication, they are currently just _wrong_ and I imagine fidelty with conversion would be better than wrong.
